hal/src/boron/network/sara_ncp_client.cpp

```diff
@@ -498,53 +498,63 @@ int SaraNcpClient::getSignalQuality(CellularSignalQuality* qual) {
-                &rscp, &ecn0, &rsrq, &rsrp));
-        CHECK_TRUE(r == 6, SYSTEM_ERROR_AT_RESPONSE_UNEXPECTED);
-        r = CHECK_PARSER(resp.readResult());
-        CHECK_TRUE(r == AtResponse::OK, SYSTEM_ERROR_AT_NOT_OK);
+        int rsrp;
+        int rsrq_n;
+        unsigned long rsrq_f;
 
-        switch (qual->strengthUnits()) {
-            case CellularStrengthUnits::RXLEV: {
-                qual->strength(rxlev);
-                break;
-            }
-            case CellularStrengthUnits::RSCP: {
-                qual->strength(rscp);
-                break;
-            }
-            case CellularStrengthUnits::RSRP: {
-                qual->strength(rsrp);
-                break;
+        // Set UCGED to mode 5 for RSRP/RSRQ values on R410M
+        CHECK_PARSER_OK(parser_.execCommand("AT+UCGED=5"));
+
+        // Default to invalid value because UCGED command does not explicitly give an ivalid value
+        qual->strength(255);
+        qual->quality(255);
+
+        auto resp = parser_.sendCommand("AT+UCGED?");
+        char buf[1024] = {};     // keep this string long in case there are multiple tower infos in the string
+        while (resp.hasNextLine()) {
+            CHECK_PARSER(resp.readLine(buf, sizeof(buf)));
+            int r = 0;
+            if (strstr(buf, "+RSRP:")) {    // check for RSRP response
+                r = sscanf(buf, "\r\n+RSRP: %*d,%*d,\"%d.%*u\"",&rsrp);
+                CHECK_TRUE(r == 1, SYSTEM_ERROR_AT_RESPONSE_UNEXPECTED);
             }
-            default: {
-                // Do nothing
-                break;
+            if (strstr(buf, "+RSRQ:")) {
+                r = sscanf(buf, "\r\n+RSRQ: %*d,%*d,\"%d.%lu\"",&rsrq_n, &rsrq_f);
+                CHECK_TRUE(r == 2, SYSTEM_ERROR_AT_RESPONSE_UNEXPECTED);
             }
+            buf[0] = '\0';
         }
 
-        switch (qual->qualityUnits()) {
-            case CellularQualityUnits::RXQUAL: {
-                qual->quality(rxqual);
-                break;
-            }
-            case CellularQualityUnits::ECN0: {
-                qual->quality(ecn0);
-                break;
-            }
-            case CellularQualityUnits::RSRQ: {
-                qual->quality(rsrq);
-                break;
-            }
-            default: {
-                // Do nothing
-                break;
-            }
+        if (rsrp < -140) {
+            qual->strength(0);
         }
-    } else {
+        else if (rsrp >= -44) {
+            qual->strength(97);
+        }
+        else {
+            qual->strength(rsrp + 140);
+        }
+
+        int rsrq_mul_100 = rsrq_n * 100 - rsrq_f;
+        if (rsrq_mul_100 < min_rsrq_mul_by_100) {
+            qual->quality(0);
+        }
+        else if (rsrq_mul_100 >= max_rsrq_mul_by_100) {
+            qual->quality(34);
+        }
+        else {
+            qual->quality((rsrq_mul_100 + 2000)/50);
+        }
+        const int r = CHECK_PARSER(resp.readResult());
```

Comment:
If one of the RSRP/RSRQ is not valid, let's return an ERROR.

Or check up above in scanf. 
**If !(r1 && r2) is one idea to invalidate this.

Reply:
^ Not doing this because if one of the RSRP/Q values is not valid, I am still collecting the other one. This coud be a useful data point to have to check why one is valid and other is not.